sm/download/src/download.rs

```diff
@@ -95,12 +111,40 @@ impl Downloader {
     }
 }
 
+fn pre_allocate_space(file: &File, file_len: u64) -> Result<(), DownloadError> {
+    let tmpstats = statvfs::statvfs("/tmp")?;
+    // Reserve 5% of total disk space
+    let five_percent_disk_space = (tmpstats.blocks() * tmpstats.block_size()) * 5 / 100;
+    let usable_disk_space =
+        tmpstats.blocks_free() * tmpstats.block_size() - five_percent_disk_space;
+
+    if file_len >= usable_disk_space {
+        return Err(DownloadError::InsufficientSpace);
+    }
+    // Reserve diskspace
+    fallocate(
+        file.as_raw_fd(),
+        FallocateFlags::empty(),
+        0,
+        file_len as i64,
+    )
+    .map_err(|e| DownloadError::FromNix(e))
```

Comment:
I would not fail the download if `fallocate` fails

Reply:
ok, ignoring the errors